### PR TITLE
net: nrf_provisioning: Add events to device mode callback

### PR DESCRIPTION
**Pull request by juhaylinen**
Update callback to send an event when provisioning state is changed.

### CODE REVIEW

**Review by jayteemo — APPROVED** [review]
this will need a change log entry.
does the nrf_provisioning rst doc need updated?

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c [hunk 1]
```diff
@@ -574,8 +577,9 @@ int nrf_provisioning_req(void)
 				LOG_DBG("Saving the latest command id");
 				commit_latest_cmd_id();
 			}
-			dm.cb(dm.user_data);
+			dm.cb(NRF_PROVISIONING_EVENT_STOP, dm.user_data);
 		}
+		dm.cb(NRF_PROVISIONING_EVENT_DONE, dm.user_data);
```
Comment (SeppoTakalo) [inline subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:582]: This looks like now that the `STOP` event is only coming if the connection was succesful.

I think the START-STOP should come in pairs.. They must be indicated regardless of the outcome.

The `DONE` should only be indicated if there was new data.

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c [hunk 2]
```diff
@@ -560,21 +563,26 @@ int nrf_provisioning_req(void)
...
 			/* Provisioning finished */
 			if (IS_ENABLED(CONFIG_NRF_PROVISIONING_SAVE_CMD_ID)) {
 				LOG_DBG("Saving the latest command id");
 				commit_latest_cmd_id();
 			}
-			dm.cb(dm.user_data);
+			dm.cb(NRF_PROVISIONING_EVENT_DONE, dm.user_data);
+		} else {
+			dm.cb(NRF_PROVISIONING_EVENT_STOP, dm.user_data);
 		}
```
Comment (SeppoTakalo) [inline subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:587]: This is still not what we agreed.. STOP must follow the START unconditionally. Just to indicate that network activities have stopped.

So it should be
```c
		dm.cb(NRF_PROVISIONING_EVENT_START, dm.user_data);
		if (IS_ENABLED(CONFIG_NRF_PROVISIONING_HTTP)) {
			ret = nrf_provisioning_http_req(&rest_ctx);
		} else {
			ret = nrf_provisioning_coap_req(&coap_ctx);
		}
		dm.cb(NRF_PROVISIONING_EVENT_STOP, dm.user_data);
```

Then `dm.cb(NRF_PROVISIONING_EVENT_DONE, dm.user_data);` should only come if there was provisioning.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c (hunk 2 above)
Comment (juhaylinen) [reply subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:584]: Oops, sorry my bad. Fixed the events.

**Review by SeppoTakalo — APPROVED** [review]

File: doc/nrf/libraries/networking/nrf_provisioning.rst [hunk 3]
```diff
@@ -65,8 +65,8 @@ The provisioning client can be initialized in one of the following ways:
 
 The function uses the following arguments:
 
-*  A pointer to a callback function, which is called when the modem state changes.
-*  A pointer that must be called once provisioning is done.
+*  A pointer to a :c:struct:`nrf_provisioning_mm_change` which holds a callback function to be called when the modem state changes.
+*  A pointer to a :c:struct:`nrf_provisioning_dm_change` which holds a callback function to be called when the provisioning state changes.
```
Comment (peknis) [inline doc/nrf/libraries/networking/nrf_provisioning.rst:69]:
```suggestion
*  A pointer to a :c:struct:`nrf_provisioning_mm_change` structure that holds a callback function to be called when the modem state changes.
*  A pointer to a :c:struct:`nrf_provisioning_dm_change` structure that holds a callback function to be called when the provisioning state changes.
```

File: include/net/nrf_provisioning.h [hunk 4]
```diff
@@ -49,23 +49,38 @@ struct nrf_provisioning_mm_change {
...
 /**
  * @struct nrf_provisioning_dm_change
- * @brief Holds the callback to be called once provisioning is done together with data set by the
- * callback provider.
+ * @brief Holds the callback to be called once provisioning state changes together with data
+ * set by the callback provider.
  *
  * @param cb        The callback function.
- * @param user_data App specific data to be fed to the callback once it's called.
+ * @param user_data Application specific data to be fed to the callback once it's called.
```
Comment (peknis) [inline include/net/nrf_provisioning.h:80]:
```suggestion
 * @param user_data Application-specific data to be fed to the callback once it is called.
```

File: include/net/nrf_provisioning.h [hunk 5]
```diff
@@ -49,23 +49,38 @@ struct nrf_provisioning_mm_change {
...
+	/** Provisioning complete. "Finished" command received from the provisioning service. */
+	NRF_PROVISIONING_EVENT_DONE
+};
+
+/**
+ * @typedef nrf_provisioning_event_cb_t
+ * @brief Called when provisioning state changes.
+ *
+ * @param event nrf_provisioning event code.
+ * @param user_data Application specific data.
```
Comment (peknis) [inline include/net/nrf_provisioning.h:70]:
```suggestion
 * @param user_data Application-specific data.
```

Comment (juhaylinen) [reply include/net/nrf_provisioning.h]: @peknis Thanks for the review. I have addressed the findings.

**Review by peknis — APPROVED** [review]

**Pull request merged**